core/trino-main/src/main/java/io/trino/operator/output/columnar/ColumnarPagePartitioner.java

```diff
@@ -0,0 +1,538 @@
+        if (replicatesAnyRow && !hasAnyRowBeenReplicated) {
+            for (IntList partitionPosition : partitionPositions) {
+                partitionPosition.add(0);
+            }
+            hasAnyRowBeenReplicated = true;
+            position = 1;
+        }
+        else {
+            position = 0;
+        }
+
+        Page partitionFunctionArgs = getPartitionFunctionArguments(page);
+
+        if (partitionFunctionArgs.getChannelCount() > 0 && onlyRleBlocks(partitionFunctionArgs)) {
+            // we need at least one Rle block since with no blocks partition function
+            // can return a different value per invocation (e.g. RoundRobinBucketFunction)
+            partitionBySingleRleValue(page, position, partitionFunctionArgs, partitionPositions);
+        }
+        else if (partitionFunctionArgs.getChannelCount() == 1 && isDictionaryProcessingFaster(partitionFunctionArgs.getBlock(0))) {
+            partitionBySingleDictionary(page, position, partitionFunctionArgs, partitionPositions);
+        }
+        else {
+            partitionGeneric(page, position, aPosition -> partitionFunction.getPartition(partitionFunctionArgs, aPosition), partitionPositions);
+        }
+        return partitionPositions;
+    }
+
+    private void appendToOutputPartition(PageBuilder outputPartition, Page page, IntArrayList positions, PositionsAppender[] positionsAppenders)
+    {
+        outputPartition.declarePositions(positions.size());
+
+        for (int channel = 0; channel < positionsAppenders.length; channel++) {
+            Block partitionBlock = page.getBlock(channel);
+            BlockBuilder target = outputPartition.getBlockBuilder(channel);
+            positionsAppenders[channel].appendTo(positions, partitionBlock, target);
+        }
+    }
+
+    private IntArrayList[] initPositions(Page page)
+    {
+        // We allocate new arrays for every page (instead of caching them) because we don't
+        // want memory to explode in case there are input pages with many positions, where each page
+        // is assigned to a single partition entirely.
+        // For example this can happen for partition columns if they are represented by RLE blocks.
+        IntArrayList[] partitionPositions = new IntArrayList[partitionFunction.getPartitionCount()];
+        for (int i = 0; i < partitionPositions.length; i++) {
+            partitionPositions[i] = new IntArrayList(initialPartitionSize(page.getPositionCount() / partitionFunction.getPartitionCount()));
+        }
+        return partitionPositions;
+    }
+
+    private static int initialPartitionSize(int averagePositionsPerPartition)
+    {
+        // 1.1 coefficient compensates for the not perfect hash distribution.
+        // 32 compensates for the case when averagePositionsPerPartition is small,
+        // and we would see more variance in the hash distribution.
+        return (int) (averagePositionsPerPartition * 1.1) + 32;
```

Comment:
Extract those constants

Reply:
unless we are going to reuse them, its way easier to read if we have the values directly here with comments.